File: DesktopCore/src/burlton/desktopcore/code/screen/ProgressDialog.java
package burlton.desktopcore.code.screen;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JPanel;
import javax.swing.JProgressBar;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;
import javax.swing.border.EmptyBorder;

import burlton.core.code.util.Debug;

public class ProgressDialog extends JDialog
							implements ActionListener
{	
	private int workToDo = 0;
	private boolean cancelPressed = false;
	private String message = "";
	
	public ProgressDialog(int workToDo, String message) 
	{
		try
		{
			this.workToDo = workToDo;
			this.message = message;
			
			JPanel panel = new JPanel();
			panel.setBorder(new EmptyBorder(10, 0, 0, 0));
			getContentPane().add(panel, BorderLayout.CENTER);
			progressBar.setPreferredSize(new Dimension(200, 20));
			progressBar.setStringPainted(true);
			panel.add(progressBar);
			
			JPanel panel_1 = new JPanel();
			panel_1.setBorder(new EmptyBorder(5, 0, 5, 0));
			getContentPane().add(panel_1, BorderLayout.SOUTH);
			panel_1.add(btnCancel);
			setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
			
			btnCancel.addActionListener(this);
			btnCancel.setVisible(false);
		}
		catch (Throwable t)
		{
			Debug.stackTrace(t);
		}
	}
	
	private JProgressBar progressBar = new JProgressBar();
	private JButton btnCancel = new JButton("Cancel");
	
	public static ProgressDialog factory(String title, String message, int workToDo)
	{
		ProgressDialog dialog = new ProgressDialog(workToDo, message);
		dialog.resetProgress();
		dialog.setTitle(title);
		dialog.setSize(300, 90);
		dialog.setResizable(false);
		dialog.setLocationRelativeTo(null);
		dialog.setModal(true);
		
		return dialog;
	}
	
	public void setVisibleLater()
	{
		SwingUtilities.invokeLater(new Runnable()
		{
			@Override
			public void run()
			{
				setVisible(true);
			}
		});
	}
	
	public void resetProgress()
	{
		SwingUtilities.invokeLater(new Runnable()
		{
			@Override
			public void run()
			{
				progressBar.setMinimum(0);
				progressBar.setMaximum(workToDo);
				progressBar.setValue(0);
				progressBar.setString(workToDo + " " + message);
			}
		});
	}
	
	public void incrementProgressLater()
	{
		incrementProgressLater(1);
	}
	
	public void incrementProgressLater(final int increment)
	{
		SwingUtilities.invokeLater(new Runnable()
		{
			@Override
			public void run()
			{
				int newValue = progressBar.getValue() + increment;
				progressBar.setValue(newValue);
				progressBar.setString((workToDo - newValue) + " " + message);
				progressBar.repaint();
			}
		});
	}
	
	
	
	public boolean cancelPressed()
	{
		return cancelPressed;
	}
	
	public void disposeLater()
	{
		SwingUtilities.invokeLater(new Runnable()
		{
			@Override
			public void run()
			{
				dispose();
			}
		});
	}

	@Override
	public void actionPerformed(ActionEvent arg0) 
	{
		cancelPressed = true;
	}
	
	public void showCancel(boolean showCancel)
	{
		btnCancel.setVisible(showCancel);
		
		int height = showCancel?120:90;
		setSize(300, height);
	}
}

File: DesktopCore/src/burlton/dartzee/code/core/screen/ProgressDialog.kt
package burlton.dartzee.code.core.screen

import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.*
import javax.swing.border.EmptyBorder

class ProgressDialog(private var workToDo: Int, private var message: String) : JDialog(), ActionListener
{
    private var cancelPressed = false

    private val progressBar = JProgressBar()
    private val btnCancel = JButton("Cancel")

    init
    {
        val panel = JPanel()
        panel.border = EmptyBorder(10, 0, 0, 0)
        contentPane.add(panel, BorderLayout.CENTER)
        progressBar.preferredSize = Dimension(200, 20)
        progressBar.isStringPainted = true
        panel.add(progressBar)

        val panelCancel = JPanel()
        panelCancel.border = EmptyBorder(5, 0, 5, 0)
        contentPane.add(panelCancel, BorderLayout.SOUTH)
        panelCancel.add(btnCancel)
        defaultCloseOperation = WindowConstants.DO_NOTHING_ON_CLOSE

        btnCancel.addActionListener(this)
        btnCancel.isVisible = false
    }

    fun setVisibleLater()
    {
        SwingUtilities.invokeLater { isVisible = true }
    }

    fun resetProgress()
    {
        SwingUtilities.invokeLater {
            progressBar.minimum = 0
            progressBar.maximum = workToDo
            progressBar.value = 0
            progressBar.string = "$workToDo $message"
        }
    }

    fun incrementProgressLater(increment: Int = 1)
    {
        SwingUtilities.invokeLater {
            val newValue = progressBar.value + increment
            progressBar.value = newValue
            progressBar.string = (workToDo - newValue).toString() + " " + message
            progressBar.repaint()
        }
    }


    fun cancelPressed(): Boolean
    {
        return cancelPressed
    }

    fun disposeLater()
    {
        SwingUtilities.invokeLater { dispose() }
    }

    override fun actionPerformed(arg0: ActionEvent)
    {
        cancelPressed = true
    }

    fun showCancel(showCancel: Boolean)
    {
        btnCancel.isVisible = showCancel

        val height = if (showCancel) 120 else 90
        setSize(300, height)
    }

    companion object
    {
        fun factory(title: String, message: String, workToDo: Int): ProgressDialog
        {
            val dialog = ProgressDialog(workToDo, message)
            dialog.resetProgress()
            dialog.title = title
            dialog.setSize(300, 90)
            dialog.isResizable = false
            dialog.setLocationRelativeTo(null)
            dialog.isModal = true

            return dialog
        }
    }
}
